packages/berry-fslib/sources/FakeFS.ts

```diff
@@ -15,77 +17,186 @@ export type WriteFileOptions = Partial<{
+        return;
+      }
+    } catch (error) {
+      // ignore errors, no big deal
+    }
 
-  abstract rmdirPromise(p: string): Promise<void>;
-  abstract rmdirSync(p: string): void;
+    this.writeFileSync(p, content);
+  }
 
-  abstract symlinkPromise(target: string, p: string): Promise<void>;
-  abstract symlinkSync(target: string, p: string): void;
+  async movePromise(fromP: P, toP: P) {
+    try {
+      await this.renamePromise(fromP, toP);
+    } catch (error) {
+      if (error.code === `EXDEV`) {
+        await this.copyPromise(toP, fromP);
+        await this.removePromise(fromP);
+      } else {
+        throw error;
+      }
+    }
+  }
 
-  abstract renamePromise(oldP: string, newP: string): Promise<void>;
-  abstract renameSync(oldP: string, newP: string): void;
+  moveSync(fromP: P, toP: P) {
+    try {
+      this.renameSync(fromP, toP);
+    } catch (error) {
+      if (error.code === `EXDEV`) {
+        this.copySync(toP, fromP);
+        this.removeSync(fromP);
+      } else {
+        throw error;
+      }
+    }
+  }
 
-  abstract copyFilePromise(sourceP: string, destP: string, flags?: number): Promise<void>;
-  abstract copyFileSync(sourceP: string, destP: string, flags?: number): void;
+  async lockPromise(affectedPath: P, callback: () => Promise<void>) {
+    const lockPath = `${affectedPath}.lock` as P;
 
-  abstract writeFilePromise(p: string, content: string | Buffer | ArrayBuffer | DataView, opts?: WriteFileOptions): void;
-  abstract writeFileSync(p: string, content: string | Buffer | ArrayBuffer | DataView, opts?: WriteFileOptions): void;
+    const interval = 1000 / 60;
+    const timeout = Date.now() + 60 * 1000;
 
-  abstract unlinkPromise(p: string): Promise<void>;
-  abstract unlinkSync(p: string): void;
+    let fd = null;
 
-  abstract utimesPromise(p: string, atime: Date | string | number, mtime: Date | string | number): Promise<void>;
-  abstract utimesSync(p: string, atime: Date | string | number, mtime: Date | string | number): void;
+    while (fd === null) {
+      try {
+        fd = await this.openPromise(lockPath, `wx`);
+      } catch (error) {
+        if (error.code === `EEXIST`) {
+          if (Date.now() < timeout) {
+            await new Promise(resolve => setTimeout(resolve, interval));
+          } else {
+            throw new Error(`Couldn't acquire a lock in a reasonable time (via ${lockPath})`);
+          }
+        } else {
+          throw error;
+        }
+      }
+    }
 
-  abstract readFilePromise(p: string, encoding: 'utf8'): Promise<string>;
-  abstract readFilePromise(p: string, encoding?: string): Promise<Buffer>;
+    try {
+      await callback();
+    } finally {
+      await this.closePromise(fd);
+      await this.unlinkPromise(lockPath);
+    }
+  }
+};
 
-  abstract readFileSync(p: string, encoding: 'utf8'): string;
-  abstract readFileSync(p: string, encoding?: string): Buffer;
+export abstract class BasePortableFakeFS extends FakeFS<PortablePath> {
```

Comment:
Why are some functions implemented in `BasePortableFakeFS` (`remove`, ...) and others in `FakeFS` (`changeFile`, ...)? What's the criteria?

Reply:
That's a leftover from before I added the `pathUtils` property to `FakeFS`. The functions in `BasePortableFakeFS` all use this property, the ones in `FakeFS` don't. I'll move these back, with the obvious exception of `resolve`